[BUU] Fixes Products Page ActionView::Template::Error

chahmedejaz opened the pull request:
#### What? Why?

- Closes #12705

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->
- This PR fixes this Error by using the previously implemented logic as per https://github.com/openfoodfoundation/openfoodnetwork/blame/6d6f8735e3eff94f53a142538e158116331175cd/app/views/admin/products_v3/_variant_row.html.haml#L22


#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- Visit the Products page, and make sure that some product variants have `nil` as their `unit_value`. i.e. very old products, created at the time when we didn't have validation for `unit_value`
- You should be able to view the page without any errors.

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [x] Feature toggled

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

rioug reviewed (approved):
Looks good :+1: 
However I am wondering if we should show a blank value when `unit_value` and `unit_description` are nil ? @openfoodfoundation/core-devs ?

dacook commented:
That sounds logical to me, I forget if there's a reason we didn't before. 
I guess with this current method, any time you save the record, it will get updated to `1` without the user necessarily noticing. Essentially it defaults to 1, which might not be correct for that product.

So if we let it be blank instead, then when you save you should get an error message. Then the user can choose the appropriate number.  I think that would be better, if the interface can support it. @chahmedejaz would you be able to test it out and see what works best?

chahmedejaz commented:
> That sounds logical to me, I forget if there's a reason we didn't before. I guess with this current method, any time you save the record, it will get updated to `1` without the user necessarily noticing. Essentially it defaults to 1, which might not be correct for that product.
> 
> So if we let it be blank instead, then when you save you should get an error message. Then the user can choose the appropriate number. I think that would be better, if the interface can support it.  would you be able to test it out and see what works best?

Sure, David. Let me do a spike on it and get back to you. Thanks

dacook commented:
Hi @chahmedejaz , sorry I didn't have an answer before. 

I think it's helpful to remember this is a rare case and considered corrupt data. The error above ☝️ means the whole page fails to load which is very disruptive. But if _cloning_ a corrupt product fails, I think that's much less disruptive and easy to work around, so we don't need to worry about solving it now. (the user can just manually create a new product with the same details)

chahmedejaz commented:
> Hi @chahmedejaz , sorry I didn't have an answer before.
> 
> I think it's helpful to remember this is a rare case and considered corrupt data. The error above ☝️ means the whole page fails to load which is very disruptive. But if _cloning_ a corrupt product fails, I think that's much less disruptive and easy to work around, so we don't need to worry about solving it now. (the user can just manually create a new product with the same details)

Thanks @dacook 👍

chahmedejaz commented on app/helpers/admin/products_helper.rb:
```diff
@@ -11,19 +11,20 @@ def product_image_form_path(product)
 
     def prepare_new_variant(product)
-      product.variants.build do |variant|
-        variant.unit_value = 1.0 * (product.variant_unit_scale || 1)
```
Mimicking old UI behavior, when creating a new variant, then both of these are empty

chahmedejaz commented on app/views/admin/products_v3/_variant_row.html.haml:
```diff
@@ -18,7 +18,7 @@
       = f.hidden_field :unit_description
       = f.text_field :unit_value_with_description,
-          value: unit_value_with_description(variant), 'aria-label': t('admin.products_page.columns.unit_value'), required: true
+          value: unit_value_with_description(variant), 'aria-label': t('admin.products_page.columns.unit_value')
```
Having `required` here is causing an error upon saving in the following scenario:
- **New varaint** button is clicked to create a new variant
- As per Old UI behavior, unit_value is empty for this product.
- However, in the new UI, if this variant is changed then the saving form doesn't work because this field is hidden on the UI until its pop-up is opened
- JS tries to find it to show the required validation error but gets broken because it's not able to find it on the UI
- Hence neither the required error gets displayed nor the form gets submitted.
- This PR removes the required validation which won't cause the failure of the above use-case however, the backend is [handling this blank field error](https://github.com/openfoodfoundation/openfoodnetwork/pull/12710/files#diff-d8fe63147fa63b4c9fb35f7c510dd1c1990c61855be8420f1ce7ebf78967e2ddR517), so it will return the erorr "_can't be blank"_ which will be displayed below it.

rioug replied:
> JS tries to find it to show the required validation error but gets broken because it's not able to find it on the UI

This validation is handled by the browser, probably in Javascript. I think that's a fair trade off, I am wondering if we could show the pop up with error on form submit so we can still benefit from the browser validation but it's probably not worth the effort.

rioug replied:
FYI: I ran into the same issue while working on the Product Refactor and I managed to implement the solution I was talking about above. It ended up pretty simple in the end.

chahmedejaz replied:
Thanks for implementing this one. ♥️ 
I tried to implement the above, maybe I was looking at it from a different POV and was facing issues. Looking forward to your solution to this in the refactor 😄

dacook replied:
The `required` JS validation was added after discussing with Mario. It ensures that the popout can't be closed if it's empty, which helps avoid the problem being hidden.
But as you've found, we have a problem if the field is empty and the popout never gets opened.

This change seems like the most practical best solution 👍  
I'm not sure if the red border will show on the popout, until you open it. But if there error text is there, I think that's ok.

chahmedejaz commented on spec/system/admin/products_v3/update_spec.rb:
```diff
@@ -58,9 +58,6 @@
       end
 
       # Unit popout
-      fill_in "Unit value", with: ""
```
As we don't have required validation as per [this](https://github.com/openfoodfoundation/openfoodnetwork/pull/12710/files#r1703198190), so we can remove its expectation

dacook replied:
Can the case of blank unit value now be added here?

https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/admin/products_v3/update_spec.rb#L251

chahmedejaz commented:
@dacook @rioug - I've addressed your comments above, please review.
Here are the changes:
https://github.com/openfoodfoundation/openfoodnetwork/pull/12710/commits/1014a50aff403cc83347bd12aa9b0ee7bda73590
https://github.com/openfoodfoundation/openfoodnetwork/pull/12710/commits/d4e0b2ab51d36bd71b5dce4800221efecb679468

Thanks

rioug commented on app/helpers/admin/products_helper.rb:
```diff
@@ -11,19 +11,20 @@ def product_image_form_path(product)
     end
 
     def prepare_new_variant(product)
```
We could remove the helper has it's not doing much now, but that's fine as it is.

rioug reviewed (approved):
Looks good :+1: Thanks @chahmedejaz

dacook reviewed (changes requested):
Actually sorry I just remembered that we are losing test coverage here. I think the removed test can still be covered in a different way:

chahmedejaz commented:
>Can the case of blank unit value now be added here?

> https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/admin/products_v3/update_spec.rb#L251

Sure, David. 
Sorry, I thought the blank error functionality is already tested in the above spec and the removed spec was just validating that if the pop-up remains open if the field is blank 😅 
No worries, I'll it now 😄

dacook commented:
> Sure, David. Sorry, I thought the blank error functionality is already tested in the above spec and the removed spec was just validating that if the pop-up remains open if the field is blank 😅 No worries, I'll it now 😄

I think it is covered for `"adding variants"`, but it would be worth covering for `"updating"`>`"with invalid data"` also. That is, at `"shows errors for both product and variant fields"`. I hope that makes sense.

chahmedejaz commented:
> > Sure, David. Sorry, I thought the blank error functionality is already tested in the above spec and the removed spec was just validating that if the pop-up remains open if the field is blank 😅 No worries, I'll it now 😄
> 
> I think it is covered for `"adding variants"`, but it would be worth covering for `"updating"`>`"with invalid data"` also. That is, at `"shows errors for both product and variant fields"`. I hope that makes sense.

Sure, David 👍🏻

chahmedejaz commented:
> Can the case of blank unit value now be added here?
> https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/admin/products_v3/update_spec.rb#L251

Hi @dacook - It's fixed here: https://github.com/openfoodfoundation/openfoodnetwork/pull/12710/commits/647a384561138ce831ead2ac9ce62165d091322b. Thanks.

dacook reviewed (approved):
Great, thank you!

filipefurtad0 commented:
Hey @chahmedejaz ,

I've edited the DB to trigger the error, and could observer the error 500 when rendering the /products page.

After staging the PR, I can confirm the page renders, and that `unit_value` remains empty:

![image](https://github.com/user-attachments/assets/5da8e18a-b838-4bd6-9504-39b4427880c1)

![image](https://github.com/user-attachments/assets/5a77505d-bd76-469f-860f-048678f77cee)

![image](https://github.com/user-attachments/assets/45b04a97-6f92-4fa7-84a6-51967436b3cc)

We should keep in mind that this PR only affects the products page and that the shopfront breaks in this case:

![image](https://github.com/user-attachments/assets/30882d5d-4594-41b0-a741-c5c6e13e5362)

But at least the user has the possibility to access the products page and fix it - provided they can find the variant which is responsible.

I believe this is the intended scenario - the `/products` page does not break - I'd say this is looking good. Merging! Thanks.
